app/src/main/java/com/zeapo/pwdstore/autofill/oreo/ui/AutofillFilterActivity.kt

```diff
@@ -112,35 +116,53 @@ class AutofillFilterView : AppCompatActivity() {
-                        DirectoryStructure.DirectoryBased -> {
-                            title.text =
-                                item.file.relativeTo(item.rootDir).parentFile?.parent ?: "/INVALID"
-                            subtitle.text =
-                                Paths.get(item.file.parentFile.name, item.file.nameWithoutExtension)
-                                    .toString()
-                        }
-                    }
-                }
-                onClick { decryptAndFill(item) }
+        val searchableAdapter = DelegatedSearchableRepositoryAdapter(
+            R.layout.oreo_autofill_filter_row,
+            ::PasswordViewHolder
+        ) { item ->
+            val file = item.file.relativeTo(item.rootDir)
+            val pathToIdentifier = directoryStructure.getPathToIdentifierFor(file)
+            val identifier = directoryStructure.getIdentifierFor(file) ?: "INVALID"
+            val accountPart = directoryStructure.getAccountPartFor(file)
+            title.text = buildSpannedString {
+                pathToIdentifier?.let { append("$it/") }
+                bold { underline { append(identifier) } }
             }
+            subtitle.text = accountPart
+            itemView.setOnClickListener { decryptAndFill(item) }
+        }
+        rvPassword.apply {
+            adapter = searchableAdapter
+            layoutManager = LinearLayoutManager(context)
+            addItemDecoration(DividerItemDecoration(context, DividerItemDecoration.VERTICAL))
         }
-        rvPassword.addItemDecoration(DividerItemDecoration(this, DividerItemDecoration.VERTICAL))
 
-        search.addTextChangedListener { recursiveFilter(it.toString(), strict = false) }
-        val initialFilter =
-            formOrigin.getPrettyIdentifier(applicationContext, untrusted = false)
+        val initialFilter = formOrigin.getPrettyIdentifier(applicationContext, untrusted = false)
         search.setText(initialFilter, TextView.BufferType.EDITABLE)
-        recursiveFilter(initialFilter, strict = formOrigin is FormOrigin.Web)
+        val filterMode =
+            if (formOrigin is FormOrigin.Web) FilterMode.StrictDomain else FilterMode.Fuzzy
+        model.search(
+            initialFilter,
+            filterMode = filterMode,
+            searchMode = SearchMode.Recursive,
+            listFilesOnly = true
+        )
+        search.addTextChangedListener {
+            model.search(
+                it.toString(),
+                filterMode = FilterMode.Fuzzy,
+                searchMode = SearchMode.Recursive,
+                listFilesOnly = true
+            )
+        }
+        model.passwordItemsList.observe(
+            this,
+            Observer { list ->
+                searchableAdapter.submitList(list)
+                if ((list.isEmpty() && rvPasswordSwitcher.nextView.id == rvPasswordEmpty.id) ||
+                    (list.isNotEmpty() && rvPasswordSwitcher.nextView.id == rvPassword.id))
```

Comment:
_yikes_

explanatory comment for posterity?

Reply:
Added a comment that hopefully explains this monstrosity. Out of all the possible options, I went for the one that should lead to the least number of layout bugs.